content/influxdb/v2.2/api-guide/tutorials/client-library-starter/python.md

```diff
@@ -0,0 +1,720 @@
+                                     token=config.get('APP', 'INFLUX_TOKEN'),
+                                     org=config.get('APP', 'INFLUX_ORG'))
+    if device_id is None:
+        device_id = str(uuid4())
+    write_api = influxdb_client.write_api(write_options=SYNCHRONOUS)
+    point = Point('deviceauth') \
+        .tag("deviceId", device_id) \
+        .field('key', f'fake_auth_id_{device_id}') \
+        .field('token', f'fake_auth_token_{device_id}')
+    client_response = write_api.write(bucket=config.get('APP', 'INFLUX_BUCKET_AUTH'), record=point)
+    # write() returns None on success
+    if client_response is None:
+        return device_id
+    # Return None on failure
+    return None
+```
+
+{{% caption %}}[iot-api-python/api/devices.py](https://github.com/influxdata/iot-api-python/blob/f354941c80b6bac643ca29efe408fde1deebdc96/api/devices.py#L47){{% /caption %}}
+
+{{% /code-tab-content %}}
+{{< /code-tabs-wrapper >}}
+
+The `create_device(device_id)` function takes a _`device_id`_ and does the following to write virtual device data to `INFLUX_BUCKET`:
+
+1. Initialize `InfluxDBClient()` with `url`, `token`, and `org` values from the configuration.
+2. Initialize a `WriteAPI` instance for writing data to an InfluxDB bucket.
+3. Create a `Point`.
+4. Use `write_api.write()` to write the `Point` to a bucket.
+5. Check for failures--if the write was successful, `write_api` returns `None`.
+6. Return _`device_id`_ if successful; `None` otherwise.
+
+The function writes a point with the following elements:
+
+| Element     | Name       | Value                     |
+|:------------|:-----------|:--------------------------|
+| measurement |            | `deviceauth`              |
+| tag         | `deviceId` | device ID                 |
+| field       | `key`      | authorization ID          |
+| field       | `token`    | authorization (API) token |
+
+Next, [create the API to list devices](#create-the-api-to-list-devices).
+
+## Create the API to list devices
+
+Add the `/api/devices` API endpoint to do the following:
+
+1. [Retrieve and process devices](#retrieve-and-process-devices)
+2. Return a list of devices.
+
+### Retrieve and process devices
+
+Retrieve registered devices in `INFLUX_BUCKET_AUTH` and process the query results.
+
+{{< code-tabs-wrapper >}}
+{{% code-tabs %}}
+[Python](#python)
+{{% /code-tabs %}}
+{{% code-tab-content %}}
+
+1. Create a Flux query that gets the last row of each [series](/influxdb/v2.2/reference/glossary#series) that contains a `deviceauth` measurement.
+   The example query below returns rows that contain the `key` field (authorization ID) and excludes rows that contain a `token` field (to avoid exposing tokens to the UI).
+
+   ```js
+   // Flux query finds devices
+    from(bucket:`${INFLUX_BUCKET_AUTH}`)
+         |> range(start: 0)
+         |> filter(fn: (r) => r._measurement == "deviceauth" and r._field != "token")
+         |> last()
+   ```
```

Comment:
~~~suggestion
1. Create a Flux query that gets the last row of each [series](/influxdb/v2.2/reference/glossary#series) in the `deviceauth` measurement.
   The example query below returns rows that contain the `key` field (authorization ID).

   ```js
   // Flux query finds devices
    from(bucket:`${INFLUX_BUCKET_AUTH}`)
         |> range(start: 0)
         |> filter(fn: (r) => r._measurement == "deviceauth" and r._field == "key")
         |> last()
   ```
~~~